kubeadm: pull kubernetes-anywhere per job

neolit123 opened the pull request:
kubernetes-anywhere (k-a) now has a specific branch for use with
e2e tests and it is no longer needed to bake a k-a commit SHA into
the kubeadm image.

ref:
https://github.com/kubernetes/kubernetes-anywhere/issues/555
https://github.com/kubernetes/kubernetes/issues/66338

The following is used to pull the `kubeadm-e2e` branch of k-a:
`--repo=k8s.io/kubernetes-anywhere=kubeadm-e2e`

This removes a step where after fixing a k-a bug, the test-infra
kubeadm image has to be updated.

/area jobs
/assign @BenTheElder 
/cc @timothysc 
/cc @mohammedzee1000

neolit123 commented on images/kubeadm/runner:
```diff
@@ -21,14 +21,6 @@ if [ ! -e test-infra ]; then
   git clone https://github.com/kubernetes/test-infra
```
should this be moved to `--repo` as well eventually?

BenTheElder replied:
this should be left up to the underlying runner in the base image, when using bootstrap.py we have a runner that does some things like that. that runner is also needed for EG the bazel cache.

BenTheElder replied:
really we should just do away with this runner being the entrypoint and do any of:
- do things ahead of time, eg pre-installing tools in the image
- let bootstrap check things out
- let the testing script do things
instead of the runner / entrypoint.

neolit123 replied:
i need to look up what the base image is.

neolit123 commented on images/kubeadm/runner:
```diff
@@ -21,14 +21,6 @@ if [ ! -e test-infra ]; then
   git clone https://github.com/kubernetes/test-infra
 fi
 
-if [ ! -e kubernetes-anywhere ]; then
```
here we the `-e` check in the runner. i guess using `--repo` for the jobs instead would pull every time?
probably OK for smaller repos.

BenTheElder replied:
this doesn't really make sense, I don't think this check would be true under any circumstance

neolit123 commented on images/kubeadm/runner:
```diff
@@ -21,14 +21,6 @@ if [ ! -e test-infra ]; then
-  git -C kubernetes-anywhere checkout a26f42c716028c4ba967b2d2f3de5ed50acc51da
-fi
-
 # This is required until https://github.com/kubernetes/kubernetes-anywhere/issues/332
```
i've re-opened this issue as it still exists.

neolit123 commented on images/kubeadm/runner:
```diff
@@ -21,14 +21,6 @@ if [ ! -e test-infra ]; then
-
 # This is required until https://github.com/kubernetes/kubernetes-anywhere/issues/332
 # is implemented.
 if [ ! -e kubernetes-anywhere/phase1/gce/account.json ]; then
```
@BenTheElder you've mentioned that `--repo` would pull in `$PWD/../kubernetes-anywhere`
do i need to adjust the paths in this last part of the file to be `../kubernetes-anywhere/`?

neolit123 replied:
i also saw the discussion about not having a custom image.

i'm not exactly sure why the `ln -s` action is here and not in the kubernetes-anywhere tree, but is it possible to move the `./test-infra/jenkins/bootstrap.py` call in the job?

this can go in a separate PR too, i guess.

BenTheElder replied:
So when the init scripts hand off to the testing code the current directory will be a checkout of the first --repo flag, which is typically the repo under test. Additional repos will be adjacent in ./../$OTHER_REPO.

I'm not sure why we do anything in this image. Ideally we'd use the underlying runner in the base image only and not do anything at runtime in this image.

timothysc commented:
/lgtm

neolit123 commented:
@BenTheElder 
i've removed the code in question from the kubeadm/runner and made a k-a PR to add that code in there instead:
https://github.com/kubernetes/kubernetes-anywhere/pull/557

the runner now looks like the one for gcloud:
https://github.com/kubernetes/test-infra/blob/master/images/gcloud/runner

i still don't understand what the base image is, but for the kubeadm image i'm seeing that it installs terraform, so i'm not exactly sure how the kubeadm image can be removed.

possibly this has to happen in a separate PR.

neolit123 commented:
le bump. :eyes:

BenTheElder commented on images/kubeadm/runner:
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env bash
-# Copyright 2017 The Kubernetes Authors.
+# Copyright 2018 The Kubernetes Authors.
```
we should drop this runner entirely, we can use the entrypoint from the base image

BenTheElder commented:
Sorry I got behind on PR reviews, going through the backlog today.

> i still don't understand what the base image is, but for the kubeadm image i'm seeing that it installs terraform, so i'm not exactly sure how the kubeadm image can be removed:

We don't need to remove this image, but we should drop the runner. The base images for this go like:
images/bootstrap (generic-ish base testing image) -> images/kubekins-e2e (generic-ish e2e / build image) -> images/kubeadm (adds tools specifically for kubernetes-anywhere).

BenTheElder commented on config/jobs/kubernetes/sig-cluster-lifecycle/kubeadm.yaml:
```diff
@@ -9,6 +9,7 @@ periodics:
     - image: gcr.io/k8s-testimages/e2e-kubeadm:v20180730-c8036c3a7
       args:
       - --repo=k8s.io/kubernetes=master
+      - --repo=k8s.io/kubernetes-anywhere=kubeadm-e2e
```
skimming through this PR I don't see us doing this for other k-a jobs (the presubmits?) we should find them and do this for all k-a jobs

neolit123 commented:
updated:
- removed the kubeadm runner
- added `--repo` 3 times for k-a in `config/jobs/kubernetes/sig-testing/bazel-build-test.yaml` as the periodics seem to be there now (?)

BenTheElder commented:
Thanks for tackling this! :-)

neolit123 commented:
thanks a lot.
hope the tests go :green_book: after this.

BenTheElder commented:
alright let's merge this, I'll send out an image bump PR to go with it
/hold cancel

neolit123 commented:
@BenTheElder 

i think this might be the first indication that something is not right about the new k-a setup:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-kubeadm-gce-upgrade-1-10-1-11/961

> /workspace/kubernetes-anywhere/.config
perhaps it's not cloning in the correct folder.

https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-all#gce-kubeadm-upgrade-1.10-1.11

waiting on some of the other tests to start too.

BenTheElder commented:
I don't see a .config in that repo? https://github.com/kubernetes/kubernetes-anywhere/tree/kubeadm-e2e

neolit123 commented:
i don't enough details.
should it be added and if so what should it contain?

BenTheElder commented:
not sure, but `/workspace/kubernetes-anywhere/.config`, `/workspace/kubernetes-anywhere` should actually have the k-a checkout IIRC, and I don't know why it would have a .config already :thinking:
